### PR TITLE
Support for future select types

### PR DESCRIPTION
**Pull request by VelvetThePanda**
# Summary
Support for type 5-8.

- This feature is not currently in open beta
- Will brick the entire channel on android

### CODE REVIEW

**Comment by stepech** [comment]
It's time!!! ❤️ 😃

**Comment by VelvetThePanda** [comment]
That it is.

**Comment by VelvetThePanda** [comment]
@ExaInsanity and @Vamplay, when you get the chance.

File: DSharpPlus/Enums/Interaction/ComponentType.cs [hunk 1]
```diff
@@ -36,12 +36,28 @@ public enum ComponentType
         /// </summary>
         Button = 2,
         /// <summary>
-        /// A select menu.
+        /// A select menu consisting of options.
         /// </summary>
-        Select = 3,
+        StringSelect = 3,
```
Comment (ExaInsanity) [inline DSharpPlus/Enums/Interaction/ComponentType.cs:41]: breaking change, so we might want to consider keeping the old name.

alternatively we can just release 4.3.0 with a breaking change if we announce it properly, that also works.

also, the documentation change is quite redundant and should be clarified.

File: DSharpPlus/Net/Serialization/DiscordComponentJsonConverter.cs [hunk 2]
```diff
@@ -48,7 +48,7 @@ public override object ReadJson(JsonReader reader, Type objectType, object exist
                 ComponentType.ActionRow => new DiscordActionRowComponent(),
                 ComponentType.Button when (int)job["style"] is 5 => new DiscordLinkButtonComponent(),
                 ComponentType.Button => new DiscordButtonComponent(),
-                ComponentType.Select => new DiscordSelectComponent(),
+                ComponentType.StringSelect => new DiscordSelectComponent(),
```
Comment (ExaInsanity) [inline DSharpPlus/Net/Serialization/DiscordComponentJsonConverter.cs:51]: should the other select types not also be added here?

File: DSharpPlus/Clients/DiscordClient.Dispatch.cs [hunk 3]
```diff
@@ -539,7 +539,7 @@ internal async Task OnReadyEventAsync(ReadyPayload ready, JArray rawGuilds, JArr
                 channel.Discord = this;
 
                 //xdc._recipients =
-                //    .Select(xtu => this.InternalGetCachedUser(xtu.Id) ?? new DiscordUser(xtu) { Discord = this })
+                //    .StringSelect(xtu => this.InternalGetCachedUser(xtu.Id) ?? new DiscordUser(xtu) { Discord = this })
```
Comment (ExaInsanity) [inline DSharpPlus/Clients/DiscordClient.Dispatch.cs:542]: as above

**Review by ExaInsanity — COMMENTED** [review]
left a couple of notes and a few questions. if those questions are fine, it's overall fine from what i can see; though i also haven't tested it.

File: DSharpPlus/Enums/Interaction/ComponentType.cs (hunk 1 above)
Comment (Naamloos) [reply DSharpPlus/Enums/Interaction/ComponentType.cs:41]: Keep it compatible.

**Review by Naamloos — CHANGES_REQUESTED** [review]
Looks great, but I'd personally say it's better to keep `Select` as `Select` and not rename it `StringSelect`. I think this property is implied by the fact that there's selects for every entity under the sun now.

File: DSharpPlus/Net/Serialization/DiscordComponentJsonConverter.cs (hunk 2 above)
Comment (Naamloos) [reply DSharpPlus/Net/Serialization/DiscordComponentJsonConverter.cs:51]: correct

File: DSharpPlus/Enums/Interaction/ComponentType.cs (hunk 1 above)
Comment (VelvetThePanda) [reply DSharpPlus/Enums/Interaction/ComponentType.cs:41]: Compatible, but incorrect. Furthermore, the last stable release (4.2.0) was released 6 months ago (04/15).

File: DSharpPlus/Net/Serialization/DiscordComponentJsonConverter.cs (hunk 2 above)
Comment (VelvetThePanda) [reply DSharpPlus/Net/Serialization/DiscordComponentJsonConverter.cs:51]: It was a small oversight.

**Comment by VelvetThePanda** [comment]
I disagree, and for the opposite reason. Now that there's *discrimination*, the vagueness of `Select` becomes *ambiguity*. Granted, this enum is only used internally, but it could be somewhat confusing to anyone looking to make a component-based PR.

**Review by Naamloos — APPROVED** [review]

**Comment by ExaInsanity** [comment]
you know your destination if it breaks

**Comment by VelvetThePanda** [comment]
Joke's on you I stole Emzi's shotgun

**Pull request merged**